### DISCUSSION TITLE
Access client IP address in server action

### DISCUSSION THREAD

**Discussion by oxc** [vercel/next.js#49730] [App Router] [answered]
I've been experimenting with server actions, since I like the progressive enhancement approach.

For auditing reasons, I need to access the IP address of the client making the request. 

I haven't seem able to find a way to access this, is it possible? If I'm not mistaken, the function only gets passed the formData, and nothing else.

**Comment by philosophocat** [comment]
Usually, there is a proxy server in front of the node application. You can set header there (for example, proxy_set_header for nginx and read `x-real-ip`)

**Reply by oxc** [reply]
Thanks, this would indeed have been sufficient for me. I had tried accessing the headers, but had failed due to ESM-related issues (headers() only finds the requestAsyncStorage if imported as 'next/headers', but not as 'next/headers.js' which I tried as a workaround. I'm now using moduleResolution=bundler thanks to a monkey-patched next.js)
This works now, so I guess this particular problem is solved for me :)
In the end, I would think though that there should be a way to fully access the request, if needed.

**Comment by trevorwhealy** [comment]
Having this problem as well. Thought that this would do the trick, but to no avail. 

```
"use server";
import { headers } from "next/headers";
export async function getIPAddress() {
    return headers().get("x-forwarded-for");
}
```

I guess the downside of the simplification of server actions is some loss of advanced functionality like this.

Would be nice if we could somehow access the standard req and res objects that are available on typical routes.

**Reply by antoinerousseau** [reply]
this method works for me! maybe it got recently added

**Comment by leerob** [comment, marked as answer]
Updated the docs here to make it more clear this depends on the infra provider, but you can also read some things from headers if provided: https://github.com/vercel/next.js/pull/59719

**Comment by ryanhs** [comment]
i forked request-ip  for nextjs / Headers friendly

https://www.npmjs.com/package/next-request-ip


```javascript
import { getClientIp } from "next-request-ip";
import { headers } from "next/headers";

export default async function Page() {
  const headersList = await headers();
  const clientIp = getClientIp(headersList);

  return (
    <div>
      <p>Your IP: {clientIp}</p>
    </div>
  );
}
```
